### ISSUE TITLE
Make Matomo PHP 7.3 compatible

### ISSUE THREAD

**Issue by Findus23** [matomo-org/matomo#13418]
In 3 months PHP 7.3 will be released, so let's start making Matomo compatible!

I tried setting up a new Matomo instance using PHP 7.3 RC2 (so all bugs found now will probably also be in the final release) and most things are working fine. These errors occurred:

-------
```
ErrorException: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior
#17 vendor/tedivm/jshrink/src/JShrink/Minifier.php(196): handleError
```
[Stack Trace](https://sentry.lw1.at/share/issue/d2252951764d49d5a009409bdd400202/)
from https://github.com/tedious/JShrink
 - [x] fixed
----------
```
2018/09/12 20:46:17 [error] 8919#8919: *137700 FastCGI sent in stderr: 
"PHP message: PHP Warning:  
"continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"?
in /var/www/matomo-beta/vendor/tedivm/jshrink/src/JShrink/Minifier.php on line 243"
```
also from https://github.com/tedious/JShrink
 - [x] fixed
--------
```
WARNING [2018-09-12 19:04:26] 20837  /var/www/matomo-beta/plugins/LanguagesManager/Commands/CreatePull.php(217): Warning - "continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? - Matomo 3.6.0 - Please report this message in the Matomo forums: https://forum.matomo.org (please do a search first as it might have been reported already)
```
https://github.com/matomo-org/matomo/blob/b8e4d27f2151103df826e5dd643b433cdde05aaf/plugins/LanguagesManager/Commands/CreatePull.php#L214-L217
 - [x] fixed in #13635
-------

```
ErrorException: compact(): Undefined variable: idGoal
#15 plugins/API/API.php(401): handleError
```
[Stack Trace](https://sentry.lw1.at/share/issue/021ddebfb19649d984b9e288f73f39bc/)
 - [x] fixed by #13593
--------

```
ErrorException: Undefined index: never
#15 core/View/HtmlReportEmailHeaderView.php(41): handleError
#14 core/View/HtmlReportEmailHeaderView.php(41): __construct
#13 core/ReportRenderer/Html.php(84): renderFrontPage
```
[Stack Trace](https://sentry.lw1.at/share/issue/92d331022a934ad688be6f05af5eb364/)
 - [x] moved to #13638
--------
To be continued...

If you want to, I can create test-admin accounts on the php 7.3 instance.

**Comment by fdellwing** [comment]
> During visitorgenerator:generate-visits. I am not sure what exactly is the error and if it has anything to do with PHP 7.3.

Not a PHP error, it tries to write -5 into an db field that is unsigned mostlikely. Why it tries to write -5 is a interesting question nethertheless.

**Comment by Findus23** [comment]
Okay, I have now updated the server to the latest release candidate and tried around and couldn't really find more issues. 

For the first two I'll create a pull request soon once https://github.com/tedious/JShrink/issues/80 is fixed.

@fdellwing I agree that this is probably a really obscure visitor generator bug, so I have removed it here.

Interestingly everything around PHP generation still works fine (where I expected most bugs).

As PHP7.3 just migrated to Debian testing, I'll expect it to be the default version in the next Debian stable (will be released in about a year) and ubuntu versions. Therefore probably a lot of people will be using it in a year.

**Comment by Findus23** [comment]
Okay, I mixed a bit up regarding JShirk. We are using a more than 4 year old beta version of it. 
The latest version from last month fixes the warnings in PHP 7.3, so it should be easy to update, but unfortunately the recent versions dropped support for 5.5 which causes composer to refuse installing it.

Does anyone have an idea what's the easiest way to fix this?

**Comment by sgiehl** [comment]
If the current version is actually still compatible with 5.5, but the dependencies were just changed in the composer.json we could overwrite it with a custom package in composer.json to overwrite the php requirement. e.g.:
```
{
    "type": "package",
    "package": {
        "name": "tedious/jshrink",
        "description": "Javascript Minifier built in PHP",
        "keywords": ["minifier","javascript"],
        "homepage": "http://github.com/tedious/JShrink",
        "type": "library",
        "license": "BSD-3-Clause",
        "version": "v1.3.1",
        "authors": [
            {
                "name": "Robert Hafner",
                "email": "tedivm@tedivm.com"
            }
        ],
        "require": {
            "php": ">=5.5.0"
        },
        "autoload": {
            "psr-0": {"JShrink": "src/"}
        },
        "source": {
            "type": "git",
            "url": "https://github.com/tedious/JShrink",
            "reference": "v1.3.1"
        }
    }
}
```

**Comment by Findus23** [comment]
@sgiehl Thanks for the tip.
I had to rename the package so that composer would use it instead of the real one.

**Comment by sgiehl** [comment]
Got another one:
```
"continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"?
/home/travis/build/piwik/vendor/tecnickcom/tcpdf/tcpdf.php:17778
```

Need to check if that will be fixed with https://github.com/matomo-org/matomo/pull/13566

**Comment by J0WI** [comment]
#11936 is still open. Shouldn't this be resolved first?

**Comment by J0WI** [comment]
Please also clarify this in the docs. Currently it just says "PHP version 5.5.9 or greater".

**Comment by sebash1992** [comment]
@sgiehl  could you do any hot fix for the tcpdf issue?

**Comment by sgiehl** [comment]
@sebash1992 this should be fixed in the upcoming 3.8.0

**Comment by socialmecpl** [comment]
> In 3 months PHP 7.3 will be released, so let's start making Matomo compatible!
> 
> I tried setting up a new Matomo instance using PHP 7.3 RC2 (so all bugs found now will probably also be in the final release) and most things are working fine. These errors occurred:
> 
> ```
> ErrorException: strpos(): Non-string needles will be interpreted as strings in the future. Use an explicit chr() call to preserve the current behavior
> #17 vendor/tedivm/jshrink/src/JShrink/Minifier.php(196): handleError
> ```
> [Stack Trace](https://sentry.lw1.at/share/issue/d2252951764d49d5a009409bdd400202/)
> from https://github.com/tedious/JShrink
> 
> * [ ]  fixed
> 
> ```
> 2018/09/12 20:46:17 [error] 8919#8919: *137700 FastCGI sent in stderr: 
> "PHP message: PHP Warning:  
> "continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"?
> in /var/www/matomo-beta/vendor/tedivm/jshrink/src/JShrink/Minifier.php on line 243"
> ```
> also from https://github.com/tedious/JShrink
> 
> * [ ]  fixed
> 
> ```
> WARNING [2018-09-12 19:04:26] 20837  /var/www/matomo-beta/plugins/LanguagesManager/Commands/CreatePull.php(217): Warning - "continue" targeting switch is equivalent to "break". Did you mean to use "continue 2"? - Matomo 3.6.0 - Please report this message in the Matomo forums: https://forum.matomo.org (please do a search first as it might have been reported already)
> ```
> [matomo/plugins/LanguagesManager/Commands/CreatePull.php](https://github.com/matomo-org/matomo/blob/b8e4d27f2151103df826e5dd643b433cdde05aaf/plugins/LanguagesManager/Commands/CreatePull.php#L214-L217)
> 
> Lines 214 to 217 in [b8e4d27](/matomo-org/matomo/commit/b8e4d27f2151103df826e5dd643b433cdde05aaf)
> 
>              switch ($returnCode) { 
>                  case 401: 
>                      $output->writeln("Pull request failed. Bad credentials... Please try again"); 
>                      continue; 
> * [x]  fixed in #13635
> 
> ```


Replace continue with break resolved the issue.

**Comment by Findus23** [comment]
@diosmosis If no one else reports any more issues, I'd say this is done.
